stdlib/Artifacts/src/Artifacts.jl

```diff
@@ -0,0 +1,634 @@
+    end
+end
+
+"""
+    ARTIFACT_OVERRIDES
+
+Artifact locations can be overridden by writing `Override.toml` files within the artifact
+directories of Pkg depots.  For example, in the default depot `~/.julia`, one may create
+a `~/.julia/artifacts/Override.toml` file with the following contents:
+
+    78f35e74ff113f02274ce60dab6e92b4546ef806 = "/path/to/replacement"
+    c76f8cda85f83a06d17de6c57aabf9e294eb2537 = "fb886e813a4aed4147d5979fcdf27457d20aa35d"
+
+    [d57dbccd-ca19-4d82-b9b8-9d660942965b]
+    c_simple = "/path/to/c_simple_dir"
+    libfoo = "fb886e813a4aed4147d5979fcdf27457d20aa35d""
+
+This file defines four overrides; two which override specific artifacts identified
+through their content hashes, two which override artifacts based on their bound names
+within a particular package's UUID.  In both cases, there are two different targets of
+the override: overriding to an on-disk location through an absolutet path, and
+overriding to another artifact by its content-hash.
+"""
+const ARTIFACT_OVERRIDES = Ref{Union{Dict{Symbol,Any},Nothing}}(nothing)
+function load_overrides(;force::Bool = false)
+    if ARTIFACT_OVERRIDES[] !== nothing && !force
+        return ARTIFACT_OVERRIDES[]
+    end
+
+    # We organize our artifact location overrides into two camps:
+    #  - overrides per UUID with artifact names mapped to a new location
+    #  - overrides per hash, mapped to a new location.
+    #
+    # Overrides per UUID/bound name are intercepted upon Artifacts.toml load, and new
+    # entries within the "hash" overrides are generated on-the-fly.  Thus, all redirects
+    # mechanisticly happen through the "hash" overrides.
+    overrides = Dict{Symbol,Any}(
+        # Overrides by UUID
+        :UUID => Dict{Base.UUID,Dict{String,Union{String,SHA1}}}(),
+
+        # Overrides by hash
+        :hash => Dict{SHA1,Union{String,SHA1}}(),
+    )
+
+    for override_file in reverse(artifacts_dirs("Overrides.toml"))
+        !isfile(override_file) && continue
+
+        # Load the toml file
+        depot_override_dict = parse_toml(override_file)
+
+        function parse_mapping(mapping::String, name::String)
+            if !isabspath(mapping) && !isempty(mapping)
+                try
+                    mapping = Base.SHA1(mapping)
+                catch e
+                    @error("Invalid override in '$(override_file)': entry '$(name)' must map to an absolute path or SHA1 hash!")
+                    rethrow()
+                end
+            end
+            return mapping
+        end
+        function parse_mapping(mapping::Dict, name::String)
+            return Dict(k => parse_mapping(v, name) for (k, v) in mapping)
+        end
+
+        for (k, mapping) in depot_override_dict
+            # First, parse the mapping. Is it an absolute path, a valid SHA1-hash, or neither?
+            try
+                mapping = parse_mapping(mapping, k)
+            catch
```

Comment:
What are the exceptions that can be thrown here? If it is only about the `Base.SHA1` call then I think it makes sense to return `nothing` from `parse_mapping` and switch on that instead of a catch all `catch` that would just give you a log message even if someone later made a typo inside `parse_mapping`. This exact thing happened at https://github.com/JuliaLang/Pkg.jl/pull/1993.